File: ios/RNZendrive/Source/InfoExtensions.swift
//
//  File.swift
//  RNZendrive
//
//  Created by Madhusudhan Sambojhu on 23/07/19.
//

import Foundation
import ZendriveSDK

protocol RnObject {
    func toRNObject() -> RnJsonObject
}

extension ZendriveSettingsError: RnObject {
    private func getErrorType() -> String {
        switch self.errorType {
        case .activityPermissionNotAuthorized:
            return "activity-recognition-permission-denied"
        case .locationPermissionNotAuthorized:
            return "location-permission-denied"
        case .locationAccuracyAuthorizationReduced:
            return "location-accuracy-authorization-reduced"
        case .locationServiceOff:
            return "location-service-off"
        @unknown default:
            return ""
        }
    }

    func toRNObject() -> RnJsonObject {
        return [
            "type": getErrorType()
        ]
    }
}

extension ZendriveSettings: RnObject {

    private func rnErrors() -> [RnJsonObject] {
        return self.errors.map { (error) in
            error.toRNObject()
        }
    }

    func toRNObject() -> RnJsonObject {
        return [
            "errors": rnErrors(),
            "warnings": []
        ]
    }

    func toErrorsFoundRNObject() -> RnJsonObject {
        return [
            "errorsFound": self.errors.count > 0 ? true : false,
            "warningsFound": false
        ]
    }
}

extension ZendriveDriveStartInfo: RnObject {
    func toRNObject() -> RnJsonObject {
        let locationPoints = waypoints as? [ZendriveLocationPoint]
        return [
            "startLocation": locationPoints?.first?.toRNObject() ?? nil,
            "driveId": driveId,
            "insurancePeriod": Utils.insurancePeriod(of: insurancePeriod),
            "sessionId": sessionId ?? nil,
            "startTimeMillis": startTimestamp,
            "trackingId": trackingId
        ]
    }
}

extension ZendriveDriveResumeInfo: RnObject {
    func toRNObject() -> RnJsonObject {
        return [
            "driveGapEndTimestampMillis": driveGapEndTimestampMillis,
            "driveGapStartTimestampMillis": driveGapStartTimestampMillis,
            "driveId": driveId,
            "insurancePeriod": Utils.insurancePeriod(of: insurancePeriod),
            "sessionId": sessionId ?? nil,
            "startTimeMillis": startTimestamp,
            "trackingId": trackingId
        ]
    }
}

extension ZendriveExtrapolationDetails: RnObject {
    convenience init(withRnObject conf: RnJsonObject) {
        self.init()
        if let num = conf["extrapolatedDistance"] as? Double {
            extrapolatedDistance = num
        }
        if let obj = conf["estimatedStartLocation"] as? RnJsonObject {
            estimatedStartLocation = ZendriveLocationPoint(
                timestamp: obj["timestampMillis"] as? Int64 ?? 0,
                latitude: obj["latitude"] as? Double ?? 0.0,
                longitude: obj["longitude"] as? Double ?? 0.0
            )
        }
    }
    func toRNObject() -> RnJsonObject {
        return [
            "extrapolatedDistance": extrapolatedDistance,
            "estimatedStartLocation": estimatedStartLocation.toRNObject()
        ]
    }
}

extension ZendriveVehicleTaggingDetails: RnObject {
    convenience init(withRnObject conf: RnJsonObject) {
        self.init()
        if let str = conf["vehicleId"] as? String {
            vehicleId = str
        }
        if let beaconResult = conf["isTaggedByBeacon"] as? Bool {
            isTaggedByBeacon = beaconResult
        }
        if let btStereoResult = conf["isTaggedByBluetoothStereo"] as? Bool {
            isTaggedByBluetoothStereo = btStereoResult
        }
    }
    func toRNObject() -> RnJsonObject {
        return [
            "vehicleId": vehicleId,
            "isTaggedByBeacon": isTaggedByBeacon,
            "isTaggedByBluetoothStereo": isTaggedByBluetoothStereo
        ]
    }
}
extension ZendriveDriveInfo: RnObject {
    convenience init(withRnObject conf: RnJsonObject) {
        self.init()
        // swiftlint:disable force_cast
        if let num = conf["averageSpeed"] as? Double {
            averageSpeed = num
        }
        if let num = conf["distanceMeters"] as? Double {
            distance = num
        }
        if let str = conf["driveId"] as? String {
            driveId = str
        }
        if let str = conf["insurancePeriod"] as? String {
            insurancePeriod = Utils.insurancePeriod(of: str)
        }
        if let str = conf["sessionId"] as? String {
            sessionId = str
        }
        if let num = conf["startTimeMillis"] as? Int64 {
            startTimestamp = num
        }
        if let num = conf["endTimeMillis"] as? Int64 {
            endTimestamp = num
        }
        if let str = conf["trackingId"] as? String {
            trackingId = str
        }
        if let obj = conf["eventRatings"] as? RnJsonObject {
            eventRatings = ZendriveEventRatings.init(rnObj: obj)
        }
        if let str = conf["driveType"] as? String {
            driveType = Utils.driveType(of: str)
        }
        if let str = conf["vehicleType"] as? String {
            vehicleType = Utils.vehicleType(of: str)
        }
        if let str = conf["userMode"] as? String {
            userMode = Utils.userMode(of: str)
        }
        if let str = conf["phonePosition"] as? String {
            phonePosition = Utils.phonePosition(of: str)
        }
        if let num = conf["maxSpeed"] as? Double {
            maxSpeed = num
        }
        if let obj = conf["score"] as? RnJsonObject {
            score = ZendriveDriveScore(
                zendriveScore: (obj["zendriveScore"] as? Int32 ?? 0)
            )
        }
        if let obj = conf["extrapolationDetails"] as? RnJsonObject {
            extrapolationDetails = ZendriveExtrapolationDetails(withRnObject: obj)
        }

        if let obj = conf["vehicleTaggingDetails"] as? RnJsonObject {
            vehicleTaggingDetails = ZendriveVehicleTaggingDetails(withRnObject: obj)
        }
        // swiftlint:enable force_cast
    }

    func toRNObject() -> RnJsonObject {
        let eventsArr = events as? [ZendriveEvent]
        return [
            "averageSpeed": averageSpeed,
            "distanceMeters": distance,
            "driveId": driveId,
            "insurancePeriod": Utils.insurancePeriod(of: insurancePeriod),
            "sessionId": sessionId ?? nil,
            "startTimeMillis": startTimestamp,
            "endTimeMillis": endTimestamp,
            "trackingId": trackingId,
            "eventRatings": eventRatings.toRNObject(),
            "events": eventsArr != nil ? Utils.toArray(items: eventsArr!) : [],
            "driveType": Utils.driveType(of: driveType),
            "vehicleType": Utils.vehicleType(of: vehicleType),
            "maxSpeed": maxSpeed,
            "phonePosition": Utils.phonePosition(of: phonePosition),
            "score": score.toRNObject(),
            "warnings": tripWarnings.map { Utils.driveInfoWarning(of: $0.tripWarningType) },
            "extrapolationDetails": extrapolationDetails?.toRNObject(),
            "userMode": Utils.userMode(of: userMode),
            // swiftlint:disable force_cast
            "waypoints": Utils.toArray(
                items: waypoints.map {
                    ZendriveLocationPointWithTimeStamp(fromLocationPoint: $0 as! ZendriveLocationPoint) }
            )
            // swiftlint:enable force_cast
        ]
    }

}

extension ZendriveActiveDriveInfo: RnObject {
    func toRNObject() -> RnJsonObject {
        return [
            "currentLocation": nil,
            "currentSpeed": currentSpeed,
            "distanceMeters": distance,
            "driveId": driveId,
            "insurancePeriod": Utils.insurancePeriod(of: insurancePeriod),
            "sessionId": sessionId ?? nil,
            "startTimeMillis": startTimestamp,
            "trackingId": trackingId
        ]
    }
}

extension ZendriveConfiguration: RnObject {
    func toRNObject() -> RnJsonObject {
        return [
            "driverId": driverId,
            "sdkKey": applicationKey,
            "managesPermissions": managesLocationPermission,
            "attributes": driverAttributes?.asDictionary() ?? [:],
            "driveDetectionMode": Utils.driveDetectionMode(of: driveDetectionMode),
            "region": Utils.region(of: region),
            "implementsMultipleAccidentCallbacks": implementsMultipleAccidentCallbacks
        ]
    }
}

extension ZendriveState: RnObject {
    func toRNObject() -> RnJsonObject {
        return [
            "isDriveInProgress": isDriveInProgress,
            "isForegroundService": false,
            "zendriveConfiguration": zendriveConfiguration.toRNObject()
        ]
    }
}

extension ZendriveLocationPoint: RnObject {
    func toRNObject() -> RnJsonObject {
        return [
            "latitude": latitude,
            "longitude": longitude,
            "timestampMillis": timestamp
        ]
    }
}

class ZendriveLocationPointWithTimeStamp: RnObject {
    var locationPoint: ZendriveLocationPoint!
    init(fromLocationPoint point: ZendriveLocationPoint) {
        locationPoint = point
    }
    func toRNObject() -> RnJsonObject {
        return [
            "location": [
                "latitude": locationPoint.latitude,
                "longitude": locationPoint.longitude
            ],
            "timestampMillis": locationPoint.timestamp
        ]
    }
}

extension ZendriveEventRatings: RnObject {
    convenience init(rnObj: RnJsonObject) {
        self.init()
        if let str = rnObj["aggressiveAccelerationRating"] as? String {
            aggressiveAccelerationRating = Utils.starRatingOf(of: str)
        }
        if let str = rnObj["hardBrakeRating"] as? String {
            hardBrakeRating = Utils.starRatingOf(of: str)
        }
        if let str = rnObj["hardTurnRating"] as? String {
            hardTurnRating = Utils.starRatingOf(of: str)
        }
        if let str = rnObj["phoneHandlingRating"] as? String {
            phoneHandlingRating = Utils.starRatingOf(of: str)
        }
        if let str = rnObj["speedingRating"] as? String {
            speedingRating = Utils.starRatingOf(of: str)
        }
    }

    func toRNObject() -> RnJsonObject {
        return [
            "aggressiveAccelerationRating": Utils.starRatingOf(of: aggressiveAccelerationRating),
            "hardBrakeRating": Utils.starRatingOf(of: hardBrakeRating),
            "hardTurnRating": Utils.starRatingOf(of: hardTurnRating),
            "phoneHandlingRating": Utils.starRatingOf(of: phoneHandlingRating),
            "speedingRating": Utils.starRatingOf(of: speedingRating)
        ]
    }
}

extension ZendriveEvent: RnObject {
    func toRNObject() -> RnJsonObject {
        return [
            "endLocation": stopLocation.toRNObject(),
            "endTimestampMillis": endTime,
            "eventType": Utils.eventType(of: eventType.rawValue),
            "severity": Utils.severity(of: eventSeverity),
            "speedingInfo": speedingData?.toRNObject(),
            "startLocation": startLocation.toRNObject(),
            "startTimestampMillis": startTime,
            "turnDirection": Utils.turnDirection(of: turnDirection)
        ]
    }
}

extension ZendriveSpeedingData: RnObject {
    func toRNObject() -> RnJsonObject {
        return [
            "avgSpeed": userSpeedMPS,
            "maxSpeed": maxUserSpeedMPS,
            "speedLimit": speedLimitMPS
        ]
    }
}

extension ZendriveDriveScore: RnObject {
    func toRNObject() -> RnJsonObject {
        return [
            "zendriveScore": zendriveScore
        ]
    }
}

extension ZendriveAccidentInfo: RnObject {
    func toRNObject() -> RnJsonObject {
        return [
            "accidentId": accidentId,
            "confidence": Utils.accidentConfidence(of: confidence),
            "driveId": driveId,
            "location": accidentLocation.toRNObject(),
            "sessionId": sessionId,
            "timestampMillis": timestamp,
            "trackingId": trackingId
        ]
    }
}

extension ZendriveVehicleInfo: RnObject {
    convenience init(withRnObject conf: RnJsonObject) {
        self.init()
        if let str =  conf["vehicleId"] as? String {
            vehicleId = str
        }
        if let str =  conf["bluetoothAddress"] as? String {
            bluetoothId = str
        }
    }

    func toRNObject() -> RnJsonObject {
        return [
            "vehicleId": vehicleId,
            "bluetoothAddress": bluetoothId
        ]
    }
}

extension ZendriveConfiguration {
    convenience init(withRnObject conf: RnJsonObject) {
        self.init()
        if let str =  conf["driverId"] as? String {
            driverId = str
        }
        if let str =  conf["sdkKey"] as? String {
            applicationKey = str
        }
        if let mode = conf["driveDetectionMode"] as? String {
            switch mode {
            case "auto-on":
                driveDetectionMode = .autoON
            case "auto-off":
                driveDetectionMode = .autoOFF
            case "insurance":
                driveDetectionMode = .insurance
            default:
                break
            }
        }
        if let attrs = conf["attributes"] as? RnJsonObject {
            driverAttributes = ZendriveDriverAttributes(withRnObject: attrs)
        }
        if let manages =  conf["managesActivityPermission"] as? Bool {
            managesActivityPermission = manages
        }
        if let manages =  conf["managesLocationPermission"] as? Bool {
            managesLocationPermission = manages
        }
        if let regionStr = conf["region"] as? String {
            switch regionStr {
            case "US":
                region = .US
            case "EU":
                region = .EU
            default:
                break
            }
        }
        if let implsMultipleCallbacks = conf["implementsMultipleAccidentCallbacks"] as? Bool {
            implementsMultipleAccidentCallbacks = implsMultipleCallbacks
        }
    }
}

extension ZendriveDriverAttributes {
    convenience init(withRnObject attrs: RnJsonObject) {
        self.init()
        if let str =  attrs["alias"] as? String {
            setAlias(str)
        }

        if let str =  attrs["groupId"] as? String {
            setGroup(str)
        }

        if let mode = attrs["serviceLevel"] as? String {
            switch mode {
            case "default":
                setServiceLevel(.levelDefault)
            case "level-1":
                setServiceLevel(.level1)
            default:
                break
            }
        }

        if let mode = attrs["vehicleType"] as? String {
            switch mode {
            case "car":
                setVehicleType(.car)
            case "motorcycle":
                setVehicleType(.motorcycle)
            default:
                break
            }
        }

        for(key, value) in attrs {
            if key != "alias" && key != "groupId" && key != "serviceLevel" && key != "vehicleType" {
                if let str =  value as? String {
                    setCustomAttribute(str, forKey: key)
                }
            }
        }
    }
}

extension ZendriveBluetoothDevice: RnObject {
    func toRNObject() -> RnJsonObject {
        [
            "name": self.name,
            "address": self.identifier
        ]
    }
}

extension ZendriveScannedBeaconInfo: RnObject {
    convenience init(withRnObject conf: RnJsonObject) {
        self.init()
        // swiftlint:disable force_cast
        if let str = conf["uuid"] as? String {
            uuid = str
        }
        if let num = conf["major"] as? Int32 {
            major = NSNumber(value: num)
        }
        if let num = conf["minor"] as? Int32 {
            minor = NSNumber(value: num)
        }
        if let num = conf["rssi"] as? Int {
            rssi = num
        }
        if let num = conf["accuracy"] as? Double {
            accuracy = num
        }
        // swiftlint:enable force_cast
    }

    func toRNObject() -> RnJsonObject {
        [
            "uuid": self.uuid,
            "major": self.major,
            "minor": self.minor,
            "rssi": self.rssi,
            "accuracy": self.accuracy
        ]
    }
}

File: android/src/main/java/com/zendrive/rn/InfoExtensions.kt
package com.zendrive.rn

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.zendrive.sdk.*
import com.zendrive.sdk.DriveInfo.Warning.UnexpectedTripDuration
import com.zendrive.sdk.PhonePosition.MOUNT
import com.zendrive.sdk.PhonePosition.UNKNOWN
import com.zendrive.sdk.ZendriveDriveType.DRIVE
import com.zendrive.sdk.ZendriveDriveType.INVALID
import com.zendrive.sdk.ZendriveDriveType.NON_DRIVING
import com.zendrive.sdk.ZendriveEvent.SpeedingInfo
import com.zendrive.sdk.ZendriveEventSeverity.HIGH
import com.zendrive.sdk.ZendriveEventSeverity.LOW
import com.zendrive.sdk.ZendriveEventType.AGGRESSIVE_ACCELERATION
import com.zendrive.sdk.ZendriveEventType.COLLISION
import com.zendrive.sdk.ZendriveEventType.HARD_BRAKE
import com.zendrive.sdk.ZendriveEventType.HARD_TURN
import com.zendrive.sdk.ZendriveEventType.PHONE_HANDLING
import com.zendrive.sdk.ZendriveEventType.PHONE_SCREEN_INTERACTION
import com.zendrive.sdk.ZendriveEventType.SPEEDING
import com.zendrive.sdk.ZendriveEventType.STOP_SIGN_VIOLATION
import com.zendrive.sdk.ZendriveEventType.HANDS_FREE_PHONE_CALL
import com.zendrive.sdk.ZendriveEventType.PASSIVE_DISTRACTION
import com.zendrive.sdk.ZendriveInsurancePeriod.Period1
import com.zendrive.sdk.ZendriveInsurancePeriod.Period2
import com.zendrive.sdk.ZendriveInsurancePeriod.Period3
import com.zendrive.sdk.ZendriveIssueType.*
import com.zendrive.sdk.ZendriveStarRating.FIVE
import com.zendrive.sdk.ZendriveStarRating.FOUR
import com.zendrive.sdk.ZendriveStarRating.NOT_AVAILABLE
import com.zendrive.sdk.ZendriveStarRating.ONE
import com.zendrive.sdk.ZendriveStarRating.THREE
import com.zendrive.sdk.ZendriveStarRating.TWO
import com.zendrive.sdk.ZendriveTurnDirection.LEFT
import com.zendrive.sdk.ZendriveTurnDirection.RIGHT
import com.zendrive.sdk.ZendriveUserMode.DRIVER
import com.zendrive.sdk.ZendriveUserMode.PASSENGER
import com.zendrive.sdk.ZendriveUserMode.UNAVAILABLE

fun ActiveDriveInfo.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  this.currentLocation?.let {
    map.putMap("currentLocation", this.currentLocation?.toRnObject())
  }
  map.putDouble("currentSpeed", this.currentSpeed)
  map.putDouble("distanceMeters", this.distanceMeters)
  map.putString("driveId", this.driveId)
  map.putString("insurancePeriod", this.insurancePeriod?.toRnObject())
  map.putDouble("startTimeMillis", this.startTimeMillis.toDouble())
  this.sessionId.let {
    map.putString("sessionId", this.sessionId)
  }
  this.trackingId.let {
    map.putString("trackingId", this.trackingId)
  }
  return map
}

fun LocationPoint.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putDouble("latitude", this.latitude)
  map.putDouble("longitude", this.longitude)
  return map
}

fun ZendriveInsurancePeriod.toRnObject(): String {
  return when (this) {
    Period1 -> "period-1"
    Period2 -> "period-2"
    Period3 -> "period-3"
  }
}

fun ZendriveEventType.toRnObject(): String {
  return when (this) {
    AGGRESSIVE_ACCELERATION -> "aggressive-acceleration"
    COLLISION -> "collision"
    HARD_BRAKE -> "hard-brake"
    HARD_TURN -> "hard-turn"
    PHONE_HANDLING -> "phone-handling"
    PHONE_SCREEN_INTERACTION -> "phone-screen-interaction"
    SPEEDING -> "speeding"
    STOP_SIGN_VIOLATION -> "stop-sign-violation"
    HANDS_FREE_PHONE_CALL -> "hands-free-phone-call"
    PASSIVE_DISTRACTION -> "passive-distraction"
  }
}

fun ZendriveExtrapolationDetails.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putMap("estimatedStartLocation", this.estimatedStartLocation?.toRnObject())
  map.putDouble("extrapolatedDistance", this.extrapolatedDistance)
  return map
}

fun ZendriveSettings.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  val errors = Arguments.createArray()
  this.errors.forEach {
    errors.pushMap(it.toRnObject())
  }
  val warnings = Arguments.createArray()
  this.warnings.forEach {
    warnings.pushMap(it.toRnObject())
  }
  map.putArray("errors", errors)
  map.putArray("warnings", warnings)
  return map
}

fun ZendriveSettingError.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("type", zendriveSettingsFromRnObject(this.type.toString()))
  return map
}

fun ZendriveSettingWarning.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("type", zendriveSettingsFromRnObject(this.type.toString()))
  return map
}

fun ZendriveIssueType.toRnObject(): String {
  return when (this) {
    POWER_SAVER_MODE_ENABLED -> "power-saver-mode-enabled"
    BACKGROUND_RESTRICTION_ENABLED -> "background-restriction-enabled"
    LOCATION_PERMISSION_DENIED -> "location-permission-denied"
    LOCATION_SETTINGS_ERROR -> "location-settings-error"
    WIFI_SCANNING_DISABLED -> "wifi-scanning-disabled"
    GOOGLE_PLAY_CONNECTION_ERROR -> "google-play-connection-error"
    GOOGLE_PLAY_SETTINGS_ERROR -> "google-play-settings-error"
    OVERLAY_PERMISSION_DENIED -> "overlay-permission-denied"
    ACTIVITY_RECOGNITION_PERMISSION_DENIED -> "activity-recognition-permission-denied"
    BATTERY_OPTIMIZATION_ENABLED -> "battery-optimization-enabled"
    ONE_PLUS_DEEP_OPTIMIZATION -> "one-plus-deep-optimization"
    AIRPLANE_MODE_ENABLED -> "airplane-mode-enabled"
    BLUETOOTH_DISABLED -> "bluetooth-disabled"
    LOCATION_UNAVAILABLE_WHILE_DRIVE_RESUME -> "location-unavailable-while-drive-resume"
    PRECISE_LOCATION_DENIED -> "precise-location-denied"
    LOCATION_MODE_HIGH_ACCURACY_DENIED -> "location-mode-high-accuracy-denied"
    BLUETOOTH_PERMISSION_DENIED -> "bluetooth-permission-denied"
    BLE_PERMISSION_DENIED -> "ble-permission-denied"
  }
}

fun ZendriveDriveType.toRnObject(): String {
  return when (this) {
    INVALID -> "invalid"
    NON_DRIVING -> "non-driving"
    DRIVE -> "drive"
  }
}

fun ZendriveVehicleType.toRnObject(): String {
  return when (this) {
    ZendriveVehicleType.CAR -> "car"
    ZendriveVehicleType.MOTORCYCLE -> "motorcycle"
  }
}

fun ZendriveVehicleTaggingDetails.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("vehicleId",this.vehicleId)
  map.putBoolean("isTaggedByBeacon",this.isTaggedByBeacon)
  map.putBoolean("isTaggedByBluetoothStereo",this.isTaggedByBluetoothStereo)
  return map
}

fun ZendriveOperationResult.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putBoolean("isSuccess", this.isSuccess)
  if (!this.isSuccess) {
    map.putString("errorCode", errorCodeString(from = this.errorCode))
    map.putString("errorMessage", this.errorMessage)
  }
  return map
}

fun ZendriveBusinessHoursOperationResult.toRnObject(): WritableMap{
  val map = Arguments.createMap()
  map.putBoolean("isSuccess", true)
  map.putString("name",this.name)
  map.putString("message", this.message)
  return map
}

fun ZendrivePauseAutoTrackingReason.toRnObject(): WritableMap{
  val map = Arguments.createMap()
  map.putString("paused_reason", this.name)
  return map
}

fun ZendriveState.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putBoolean("isDriveInProgress", this.isDriveInProgress)
  map.putBoolean("isForegroundService", this.isForegroundService)
  map.putMap("zendriveConfiguration", this.zendriveConfiguration.toRnObject())
  return map
}

fun ZendriveConfiguration.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("driverId", this.driverId)
  map.putString("sdkKey", this.sdkKey)
  map.putString(
    "driveDetectionMode",
    driveDetectionModeString(from = this.zendriveDriveDetectionMode)
  )
  map.putString(
    "region",
    regionString(from = this.region)
  )
  map.putMap("attributes", this.driverAttributes.toRnObject())
  map.putBoolean("implementsMultipleAccidentCallbacks", this.implementsMultipleAccidentCallbacks)
  return map
}

fun ZendriveDriverAttributes.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("alias", this.alias)
  map.putString("group", this.group)
  map.putString("serviceLevel", serviceLevelAttributeString(from = this.serviceLevel))
  return map
}

fun ZendriveStarRating.toRnObject(): String {
  return when (this) {
    FIVE -> "5"
    FOUR -> "4"
    THREE -> "3"
    TWO -> "2"
    ONE -> "1"
    NOT_AVAILABLE -> "not-available"
  }
}

fun ZendriveEventRatings.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("aggressiveAccelerationRating", this.aggressiveAccelerationRating.toRnObject())
  map.putString("hardBrakeRating", this.hardBrakeRating.toRnObject())
  map.putString("hardTurnRating", this.hardTurnRating.toRnObject())
  map.putString("phoneHandlingRating", this.phoneHandlingRating.toRnObject())
  map.putString("speedingRating", this.speedingRating.toRnObject())
  return map
}

fun DriveStartInfo.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("driveId", this.driveId)
  map.putString("sessionId", this.sessionId)
  map.putString("trackingId", this.trackingId)
  map.putString("insurancePeriod", this.insurancePeriod?.toRnObject())
  map.putMap("startLocation", this.startLocation?.toRnObject())
  map.putDouble("startTimeMillis", this.startTimeMillis.toDouble())
  return map
}

fun DriveResumeInfo.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("driveId", this.driveId)
  map.putString("sessionId", this.sessionId)
  map.putString("trackingId", this.trackingId)
  map.putString("insurancePeriod", this.insurancePeriod?.toRnObject())
  map.putDouble("startTimeMillis", this.startTimeMillis.toDouble())
  map.putDouble("driveGapEndTimestampMillis", this.driveGapEndTimestampMillis.toDouble())
  map.putDouble("driveGapStartTimestampMillis", this.driveGapStartTimestampMillis.toDouble())
  return map
}

fun ZendriveEventSeverity.toRnObject(): String {
  return when (this) {
    HIGH -> "high"
    LOW -> "low"
    ZendriveEventSeverity.NOT_AVAILABLE -> "not-available"
  }
}

fun SpeedingInfo.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putDouble("avgSpeed", this.avgSpeed)
  map.putDouble("maxSpeed", this.maxSpeed)
  map.putDouble("speedLimit", this.speedLimit)
  return map
}

fun ZendriveTurnDirection.toRnObject(): String {
  return when (this) {
    LEFT -> "left"
    RIGHT -> "right"
  }
}

fun PhonePosition.toRnObject(): String {
  return when (this) {
    MOUNT -> "mount"
    UNKNOWN -> "unknown"
  }
}

fun DriveScore.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putInt("zendriveScore", this.zendriveScore)
  return map
}

fun ZendriveUserMode.toRnObject(): String {
  return when (this) {
    DRIVER -> "driver"
    PASSENGER -> "passenger"
    UNAVAILABLE -> "unavailable"
  }
}

fun DriveInfo.Warning.toRnObject(): String {
  return when (this) {
    UnexpectedTripDuration -> "unexpected-trip-duration"
  }
}

fun ZendriveAccidentConfidence.toRnObject(): String {
  return when (this) {
    ZendriveAccidentConfidence.HIGH -> "high"
    ZendriveAccidentConfidence.LOW -> "low"
    ZendriveAccidentConfidence.INVALID -> "invalid"
  }
}

fun LocationPointWithTimestamp.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putDouble("timestampMillis", this.timestampMillis.toDouble())
  map.putMap("location", this.location.toRnObject())
  return map
}

fun ZendriveEvent.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putMap("endLocation", this.endLocation?.toRnObject())
  map.putDouble("endTimestampMillis", this.endTimestampMillis.toDouble())
  map.putString("eventType", this.eventType.toRnObject())
  map.putString("severity", this.severity.toRnObject())
  map.putMap("speedingInfo", this.speedingInfo?.toRnObject())
  map.putMap("startLocation", this.startLocation?.toRnObject())
  map.putDouble("startTimestampMillis", this.startTimestampMillis.toDouble())
  map.putString("turnDirection", this.turnDirection?.toRnObject())
  return map
}

fun DriveInfo.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("driveId", this.driveId)
  map.putString("vehicleType", this.vehicleType?.toRnObject())
  map.putString("sessionId", this.sessionId)
  map.putString("trackingId", this.trackingId)
  map.putString("insurancePeriod", this.insurancePeriod?.toRnObject())
  map.putDouble("startTimeMillis", this.startTimeMillis.toDouble())
  map.putDouble("averageSpeed", this.averageSpeed)
  map.putDouble("distanceMeters", this.distanceMeters)
  map.putDouble("maxSpeed", this.maxSpeed)
  map.putString("driveType", this.driveType.toRnObject())
  map.putDouble("endTimeMillis", this.endTimeMillis.toDouble())
  map.putMap("eventRatings", this.eventRatings.toRnObject())
  map.putMap("extrapolationDetails", this.extrapolationDetails?.toRnObject())
  map.putMap("vehicleTaggingDetails", this.vehicleTaggingDetails?.toRnObject())

  map.putArray(
    "events",
    this.events.fold(Arguments.createArray()) { acc, elem ->
      acc.pushMap(elem.toRnObject())
      acc
    }
  )
  map.putString("phonePosition", this.phonePosition.toRnObject())
  map.putMap("score", this.score.toRnObject())
  map.putString("userMode", this.userMode?.toRnObject())
  map.putArray(
    "warnings",
    this.warnings.fold(Arguments.createArray()) { acc, elem ->
      acc.pushString(elem.toRnObject())
      acc
    }
  )
  map.putArray(
    "waypoints",
    this.waypoints.fold(Arguments.createArray()) { acc, elem ->
      acc.pushMap(elem.toRnObject())
      acc
    }
  )
  return map
}

fun AccidentInfo.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("accidentId", this.accidentId)
  map.putString("confidence", this.confidence.toRnObject())
  map.putString("driveId", this.driveId)
  map.putString("sessionId", this.sessionId)
  map.putDouble("timestampMillis", this.timestampMillis.toDouble())
  map.putString("trackingId", this.trackingId)
  map.putMap("location", this.location?.toRnObject())
  return map
}

fun ZendriveVehicleInfo.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("vehicleId", this.vehicleId)
  map.putString("bluetoothAddress", this.bluetoothAddress)
  return map
}


fun ZendriveScannedBeaconInfo.toRnObject(): WritableMap {
  val map = Arguments.createMap()
  map.putString("uuid", this.uuid.toString())
  map.putInt("major", this.major)
  map.putInt("minor", this.minor)
  map.putInt("rssi", this.rssi)
  return map
}
